src/player.cpp

```diff
@@ -4599,6 +4599,73 @@ void player::on_hurt( Creature *source, bool disturb /*= true*/ )
     }
 }
 
+bool player::immune_to( body_part bp, damage_unit dam ) const
+{
+    // We use the worst-case scenario instead of random numbers
+    if( has_trait( "DEBUG_NODMG" ) || is_immune_damage( dam.type ) ) {
+        return true;
+    }
+    if( dam.type == DT_HEAT ) {
+        return false; // No one is immune to fire
+    }
+
+    // Bionics reduce the damage for bashing, cutting, and stabbing
+    if( has_active_bionic( "bio_ads" ) && power_level >= 25 ) {
```

Comment:
Power level check should go before bionic check because `int` comparisons are faster than `has_active_bionic` lookups.